nova/virt/vmwareapi/vmops.py

```diff
@@ -698,7 +698,47 @@ def _cache_iso_image(self, vi, tmp_image_ds_loc):
                             tmp_image_ds_loc.parent,
                             vi.cache_image_folder)
 
-    def _get_vm_config_info(self, instance, image_info, extra_specs):
+    def _get_hagroup_info(self, context, instance):
+        """Return hagroup regex and hagroup for the given instance
+
+        The hagroup computes from the server-group the instance is in and thus
+        this function may return (None, None) if the server isn't in a
+        server-group.
+        """
+        # this currently means the hagroup feature is disabled
+        if not self._datastore_hagroup_regex:
+            return None, None
+
+        instance_group_object = objects.instance_group.InstanceGroup
+        try:
+            server_group = instance_group_object.get_by_instance_uuid(
+                context, instance.uuid)
+        except nova.exception.InstanceGroupNotFound:
+            return None, None
+
+        # we don't handle affinity here, just anti-affinity
+        if 'anti-affinity' not in server_group.policy:
+            return None, None
+
+        # query all instances not deleted in server_group.members. this
+        # will only filter in the current cell, but that should be a big
+        # enough radius as cells are AZs in our env - except in qa-de-1
+        # TODO(jkulik): Do we need this to work across cells?
+        InstanceList = objects.instance.InstanceList
+        filters = {'uuid': server_group.members, 'deleted': False}
+        instances = InstanceList.get_by_filters(context, filters,
+                                                expected_attrs=[])
+        existing_instance_uuids = set(i.uuid for i in instances)
+        # we need to keep the order of the members here and thus filter the
+        # members by the found instances
+        not_deleted_members = [m for m in server_group.members
+                               if m in existing_instance_uuids]
+        member_index = not_deleted_members.index(instance.uuid)
+        hagroup = ['A', 'B'][member_index % 2]
+
+        return self._datastore_hagroup_regex, hagroup
```

Comment:
I am not quite convinced by the algorithm. I don't think, it really provides any guarantees about the distribution, so we might as well simply hash the instance-uuid. (Which has its own downsides)

AFAIK, the members in the list are ordered by the time they have been added to the list. So deleting and creating the instances in a FIFO fashion would result with all instances ending up in the hagroup 'A'.

Reply:
Yes, you're right about that. I'm up to changing this for all VMs that are not the first two. Using the hash of the instance-uuid sounds good.